### PR TITLE
ANN: detect E0317 missing else block

### PR DESCRIPTION
**Pull request by afetisov**
Also slightly change type inference for if expressions without else block.
Previously such expressions were always inferred to `()`. This is right for expression statements,
but wrong for nested expressions. We change it to infer `()` only if the parent is a block, otherwise
type inference returns the type of the `if` block.

changelog: Detect [E0317](https://doc.rust-lang.org/error-index.html#E0317), An `if` expression is missing an `else` block.

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 1]
```diff
@@ -873,6 +874,26 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
...
 
+    private fun checkIfExpr(holder: RsAnnotationHolder, expr: RsIfExpr) {
+        if (expr.elseBranch == null) {
+            val ty = if (expr.type != TyUnknown) {
+                expr.type
+            } else {
+                // If both expected and actual types are unknown, inspection is not applicable.
+                expr.expectedType ?: return
+            }
+            if (ty.isEquivalentTo(TyUnit.INSTANCE)) {
```
Comment (dima74) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:885]:
```suggestion
            if (ty is TyUnit) {
```

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 2]
```diff
@@ -873,6 +874,26 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
...
+                // If an `if` expression returns `()`, it doesn't need an `else`.
+                return
+            }
+            if (ty.isEquivalentTo(TyNever)) {
```
Comment (dima74) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:889]:
```suggestion
            if (ty == TyNever) {
```

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt [hunk 3]
```diff
@@ -1044,7 +1044,11 @@ class RsTypeInferenceWalker(
             blockTys.add(elseBranch.ifExpr?.inferType(expected))
             blockTys.add(elseBranch.block?.inferType(expected))
         }
-        return if (expr.elseBranch == null) TyUnit.INSTANCE else getMoreCompleteType(blockTys.filterNotNull())
+        return if (expr.elseBranch == null && expr.parent is RsBlock) {
```
Comment (dima74) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInferenceWalker.kt:1047]: Looks like `expr.parent is RsBlock` is always false - if `if` is inside block, its parent will be `RsExprStmt`. Also please add tests when `if` is top-level stmt or tail expr in a function

**Review by dima74 — COMMENTED** [review]
The `if` can be quite large, so lets reduce error highlighting range to some part of it, e.g. to the `if` keyword

Also can we unite this with `AddElseIntention` - that is if there is error E0317, we will provide "add else" as quick-fix for the `MissingElseBranch`, and otherwise there will be just intention "add else" as now

**Comment by afetisov** [comment]
Side note: it would be great if in-flight Github jobs could be cancelled. Afaik they make it as hard as possible to do, but maybe something [like this](https://stackoverflow.com/a/67223051) would help? I have accepted the changes and now the runners are busy with running tests which I absolutely don't care about at the moment.

**Comment by dima74** [comment]
> Side note: it would be great if in-flight Github jobs could be cancelled. Afaik they make it as hard as possible to do, but maybe something [like this](https://stackoverflow.com/a/67223051) would help? I have accepted the changes and now the runners are busy with running tests which I absolutely don't care about at the moment.

Yeah [we have this](https://github.com/intellij-rust/intellij-rust/blob/8c9ced3406f1dd82e11c2053e3f927045c4a7758/.github/workflows/check.yml#L16-L20), from my experience it works (e.g. when I push and then force push again, the first check is cancelled). The specific workflow run can be cancelled on workflow page I guess

**Comment by afetisov** [comment]
Meta: should we extract it into a separate "Missing else" inspection? I'm getting uncomfortable adding all those checks to the annotator. It is already overloaded with inspections. The test file spans 5000 (!) lines.

**Comment by afetisov** [comment]
>  we will provide "add else" as quick-fix for the MissingElseBranch

I'm not sure it's not misleading. It's not _actually_ a fix, since the branch must return some non-unit value, and we have no way to provide it automatically. It seems confusing if a proposed "fix" doesn't really fix the error. Although, it does punt a "Missing else" error into a "if and else have incompatible types" error.

Also, I'm not sure that restricting the error highlighting to `if` is useful. Firstly, a missing else is a really drastic syntax error, and providing (even empty) else branch is quite trivial, so being a bit obnoxious about it doesn't seem out of place. Secondly, rustc lint highlights the entire `if`-expression. Since the user is likely to run `cargo check` or `cargo clippy` external linter anyway, there would be no difference in terms of highlighting (but there may be confusion why a different set of errors is reported for different parts of the underlined expression).

**Comment by dima74** [comment]
> Meta: should we extract it into a separate "Missing else" inspection? I'm getting uncomfortable adding all those checks to the annotator. It is already overloaded with inspections. The test file spans 5000 (!) lines.

That's a good question. I don't know a solution at the moment. Let's extract if you want

> I'm not sure it's not misleading. It's not _actually_ a fix, since the branch must return some non-unit value, and we have no way to provide it automatically. It seems confusing if a proposed "fix" doesn't really fix the error. Although, it does punt a "Missing else" error into a "if and else have incompatible types" error.

It is normal if quick-fix will not fully fix the error, but bring it closer to the correct code. E.g. when we have a `match` with missing branches used as an expression, the quick-fix "Add remaining patterns" will add a match arm with empty body, and user will have to put some code in it

> Also, I'm not sure that restricting the error highlighting to `if` is useful. Firstly, a missing else is a really drastic syntax error, and providing (even empty) else branch is quite trivial, so being a bit obnoxious about it doesn't seem out of place. Secondly, rustc lint highlights the entire `if`-expression. Since the user is likely to run `cargo check` or `cargo clippy` external linter anyway, there would be no difference in terms of highlighting (but there may be confusion why a different set of errors is reported for different parts of the underlined expression).

Sounds reasonable, lets keep as is

**Comment by dima74** [comment]
Btw there is one regression

Also one test failed, because now we infer type of `if f { return; }` as `!` but looks like type should be `()`

**Pull request closed without merging**